### PR TITLE
Fix #3485, #3540: Don't use dropout for predicting test sets

### LINKED ISSUE #3485: Inconsistent results with dart booster
I am getting inconsistent predictions using dart booster if I save it to file and then load again.
Using other boosters I have consistent predictions and all work as should.

I am using:
- Linux 4.17.5-1-ARCH
- Python 3.6.6
- xgboost 0.72.1
- numpy 1.14.5

```
import xgboost as xgb
import numpy as np

data = np.random.rand(10,1)
labels = np.random.rand(10,1) > 0.5
train = xgb.DMatrix(data, label=labels)
predict = xgb.DMatrix(np.random.rand(1,1))

param = {'booster': 'dart', 
         'silent':'1'}
bst = xgb.Booster()
bst = xgb.train(param.items(), train)
bst.save_model("xgb_test.bin")

for i in range(0,10): #run load model and prediction ten times to see difference
    bst = xgb.Booster()
    bst.load_model('xgb_test.bin')
    new_value = bst.predict(predict)
    print(new_value)
    if i == 0:
        last_value = new_value
    if last_value != new_value:
        raise SystemExit
```

### LINKED ISSUE #3540: loading pickle model is unstable as getting inconsistent prediction (10% abs. error)
I saved the xgb regression model by pickle (protocol = 4). On loading the model by pickle and using .predict() it gives different output each time. 

training parameters : 
```
train_param = {
        'booster': 'dart',
        'sample_type': 'uniform',
        'normalize_type': 'tree',
        'rate_drop': 0.1,
        'skip_drop': 0.5,
        'objective': "reg:linear",
        'silent': False,
        'learning_rate': 0.1,
        'n_jobs': 18,
        'max_depth': 60,
        'tree_method': 'exact',
        'subsample': 0.75,
        'colsample_bytree': 0.8,
    }
```
number of trees = 30
Saving model by 
```
pickle.dump(model, open(str(model_name), "wb"),protocol=4)
```

testing : 
```
model = pickle.load(model_file.open('rb'))
print('trying batch of 5 for Predictions')
for x in range(0,5):
    print('iteration ' + str(x))
    for i in range(0,10,5):
        featureMatrix = features[i:i+5]
        originalPredictionsArray = originalPredictions[i:i+5]
        featuresDmatrix = xg.DMatrix(featureMatrix)
        predictions = model.predict(featuresDmatrix)
        for j in range(0,len(predictions)):
            percentageError = (predictions[j] - originalPredictionsArray[j])/(originalPredictionsArray[j]) * 100
            print('original : ' + str(originalPredictionsArray[j]) + ', predicted : ' + str(predictions[j]) + ', difference : ' + str(percentageError))
```
Output : 
```
trying batch of 5 for Predictions
iteration 0
original : 0.7610444, predicted : 0.7673624, difference : 0.8301770291664858
original : 2.5389864, predicted : 2.5028908, difference : -1.42165293710883
original : 2.5389864, predicted : 2.5028908, difference : -1.42165293710883
original : 0.4709874, predicted : 0.47068974, difference : -0.06319834516848555
original : 1.7682437, predicted : 1.7803591, difference : 0.6851685081183685
original : 1.7682437, predicted : 1.7321702, difference : -2.0400737641673303
original : 0.20668933, predicted : 0.22382358, difference : 8.289855681279548
original : 0.30097038, predicted : 0.31360516, difference : 4.198014276721492
original : 0.30097038, predicted : 0.31360516, difference : 4.198014276721492
original : 0.30097038, predicted : 0.31360516, difference : 4.198014276721492
iteration 1
original : 0.7610444, predicted : 0.76594466, difference : 0.6438861792991928
original : 2.5389864, predicted : 2.5410786, difference : 0.08240168221785724
original : 2.5389864, predicted : 2.5410786, difference : 0.08240168221785724
original : 0.4709874, predicted : 0.4775506, difference : 1.3934971506167761
original : 1.7682437, predicted : 1.764505, difference : -0.21143416347698926
original : 1.7682437, predicted : 1.7721883, difference : 0.22308044161545096
original : 0.20668933, predicted : 0.21596307, difference : 4.486799402870509
original : 0.30097038, predicted : 0.31220746, difference : 3.7336167112000864
original : 0.30097038, predicted : 0.31220746, difference : 3.7336167112000864
original : 0.30097038, predicted : 0.31220746, difference : 3.7336167112000864
iteration 2
original : 0.7610444, predicted : 0.73102736, difference : -3.944189756755978
original : 2.5389864, predicted : 2.3094754, difference : -9.039472527087291
original : 2.5389864, predicted : 2.3094754, difference : -9.039472527087291
original : 0.4709874, predicted : 0.47702405, difference : 1.2817006499151011
original : 1.7682437, predicted : 1.7223879, difference : -2.5932958285545586
original : 1.7682437, predicted : 1.7677094, difference : -0.030217869414977342
original : 0.20668933, predicted : 0.21680078, difference : 4.8921001893192955
original : 0.30097038, predicted : 0.3130991, difference : 4.029867086800193
original : 0.30097038, predicted : 0.3130991, difference : 4.029867086800193
original : 0.30097038, predicted : 0.3130991, difference : 4.029867086800193
iteration 3
original : 0.7610444, predicted : 0.7661159, difference : 0.6663873821777235
original : 2.5389864, predicted : 2.4622545, difference : -3.0221459937336865
original : 2.5389864, predicted : 2.4622545, difference : -3.0221459937336865
original : 0.4709874, predicted : 0.46926302, difference : -0.3661207969508767
original : 1.7682437, predicted : 1.772145, difference : 0.22063321264430102
original : 1.7682437, predicted : 1.7601038, difference : -0.4603369007079967
original : 0.20668933, predicted : 0.21152967, difference : 2.341844228651523
original : 0.30097038, predicted : 0.30832887, difference : 2.444920645884882
original : 0.30097038, predicted : 0.30832887, difference : 2.444920645884882
original : 0.30097038, predicted : 0.30832887, difference : 2.444920645884882
iteration 4
original : 0.7610444, predicted : 0.75617254, difference : -0.6401547920660151
original : 2.5389864, predicted : 2.5531857, difference : 0.559250784889555
original : 2.5389864, predicted : 2.5531857, difference : 0.559250784889555
original : 0.4709874, predicted : 0.46667275, difference : -0.9160864330435913
original : 1.7682437, predicted : 1.7390194, difference : -1.652730677287386
original : 1.7682437, predicted : 1.764505, difference : -0.21143416347698926
original : 0.20668933, predicted : 0.2056905, difference : -0.48325033497260894
original : 0.30097038, predicted : 0.285376, difference : -5.181362921731075
original : 0.30097038, predicted : 0.285376, difference : -5.181362921731075
original : 0.30097038, predicted : 0.285376, difference : -5.181362921731075
```

### PR DESCRIPTION
**Pull request by hcho3**
Dropout (for DART) should only be used at training time.

Fixes #3485.
Fixes #3540.

### CODE REVIEW

**Comment by khotilov** [comment]
@hcho3 for DART, the assumption was that for deterministic predictions, one has to call predict with ntree_limit>0. I agree that it's not the best interface design and that it is poorly documented (if any). But here you've added another layer of hidden logic on the top of that, and also these changes break the possibility of useing a custom objective (where Predict is called directly before DoBoost) with DART.

I'm sorry, I didn't have much time lately to keep up with the updates here (a newborn baby), but I will catch up at some time soon.

**Comment by hcho3** [comment]
@khotilov My reasoning was that the most common use case should be intuitive.

> But here you've added another layer of hidden logic on the top of that, and also these changes break the possibility of useing a custom objective (where Predict is called directly before DoBoost) with DART.

Let's come up with a way to accommodate this usage. We can either add a special method or add a flag to `predict()`.

**Comment by hcho3** [comment]
@khotilov Actually, let me revert the change. You're right in that we'd be breaking the custom objective:
https://github.com/dmlc/xgboost/blob/44811f233071c5805d70c287abd22b155b732727/python-package/xgboost/core.py#L1086

**Comment by RAMitchell** [comment]
In my opinion DART should be integrated with GBTree. This would require small changes to GBTree and would bring the functionality of the dart booster in line with the rest of the code. This would break serialisation and so should be done at a major release.

**Comment by khotilov** [comment]
@RAMitchell I agree
